text/entities/0264-resource-and-entities.md

```diff
@@ -0,0 +1,284 @@
+
+### How to deal with Resource/Entities whose lifecycle does not match the SDK?
+
+This proposal motivates a Resource Coordinator in the SDK whose job could include managing changes in entity lifetimes, but does not account for how these changes would be broadcast across TracerProvider, LogProvider, MeterProvider, etc.  That would be addressed in a follow on OTEP.
+
+### How to deal with Prometheus Compatibility for non-SDK telemetry?
+
+Today, Prometheus compatibility relies on two key attributes in Resource: `service.name` and `service.instance.id`. These are not guaranteed to exist outside of OpenTelemetry SDK generation. While this question is not fully answered, we believe outlining identity in all resources within OpenTelemetry allows us to define a solution in the future while preserving compatibility with what works today.
+
+A quick proposal of what this might look like:
+
+- `target_info` metric generation is updated to exclude any keys which are contained in `descriptive_attributes_keys` of an entity.
+- For each entity which has non-empty descriptive_attributes_keys, generate an info metric: `<entity_type>_entity_info` (naming TBD), which has all identifying and descriptive keys. This should play nicely with the planned improvements to [info-typed metrics](https://github.com/prometheus/proposals/blob/main/proposals/2024-04-10-native-support-for-info-metrics-metadata.md#goals).
+
+### Should entities have a domain?
+
+Is it worth having a `domain` in addition to type for entity?  We could force each entity to exist in one domain and leverage domain generically in resource management.  Entity Detectors would be responsible for an entire domain, selecting only ONE to apply a resource. Domains could be layered, e.g. a Cloud-specific domain may layer on top of a Kubernetes domain, where "GKE cluster entity" identifies *which* kubernetes cluster a kuberntes infra entity is part of.  This layer would be done naively, via automatic join of participating entities or explicit relationships derived from GKE specific hooks.
+
+It's unclear if this is needed initially, and we believe this could be layered in later.
+
+### Should resources have only one associated entity?
+
+Given the problems leading to the Entities working group, and the needs of existing Resource users today, we think it is infeasible and unscalable to limit resource to only one entity.  This would place restrictions on modeling Entities that would require OpenTelemetry to be the sole source of entity definitions and hurt building an open and extensible ecosystem.  Additionally it would need careful definition of solutions for the following problems/rubrics:
```

Comment:
You know my opinion so far is that we don't need more than one associated entity. It would be useful to list some use cases that demonstrate the need to have more than one associated entity. I would like to have a thorough debate on this topic and use cases are the necessary battleground for this debate.

Reply:
I believe you missed this debate in the Entities WG - but we did take [notes](https://docs.google.com/document/d/15Yt9ss2_EhuFPqItPbk4vjfpeRDAQ5WCUVuY_kCeOAo/edit#heading=h.y4e2fcmakmhd).  We can re-open this discussion in the meeting again.

I'll list the notes  here:

## 2024-07-18, 8am PT

```
Concerns around single vs. multi-associated
- Use case where eng domain was original modeled and had ONE associated entity
- Now, new domain is modeled, e.g. EC2
- How would we associate the *more* concrete type without breaking backward compatibility if systems rely on previously associated single-entity type.
```

Later
```
Entity Overlap - Should OTEL Signals have only ONE producing entity (via Resource)?
- Types of Entities
   - Infrastructure (Machine, VM, Container, CPU, Process)
   - Logical (Service, Application, etc.)
    - "Mixed" (Deployment, Session)
    
Example:
A k8s deployment leads to k8s pods being created on k8s nodes. Should OTLP metrics from this include K8S deployment identity *and* k8s pod identity *and* k8s node identity?
The user configured a service.name in opentelemetry.  It's also running on a virtual machine.  Do we include service.instance.id *and* vm identity in OTLP generated?
```

Additionally, from discussion on other design docs:

- [In the presence of "service" and "process" - which entity do you pick for resource?](https://docs.google.com/document/d/18FS9eFCW8N1oJdsKEQDJGFNVHPeZYx8rdCmRDTi9GNE/edit?disco=AAABLWnuOBw)
- [What about just the issue of picking between "service" or "container / process"?](https://docs.google.com/document/d/18FS9eFCW8N1oJdsKEQDJGFNVHPeZYx8rdCmRDTi9GNE/edit?disco=AAABLWnuOB8)
  - Note: I think the answer you give here needs a LOT of fleshing out how that solution would interact with the SDK and how users would be able to acheive the solution you lay out.  I don't see it as a viable thing to do in practice, and the design doesn't cover how to achieve it.
- A link in this OTEP on Resource, including [lessons from GCP's own single-resource model](https://docs.google.com/document/d/1Xd1JP7eNhRpdz1RIBLeA1_4UYPRJaouloAYqldCeNSc/edit#heading=h.1yfz555jem12)

And some offline discussions not recorded:

- Anna in the Entities WG had some concerns that I think weren't written down, I'll let her comment on those here.
- There's a use case where we need to allow the open ecosystem to define entity types. If a user is leveraging OpenTelemetry and all entity types *MUST* be defined within semconv this is problematic.  Take the following scenario.
  - User X is using a tool that leveraged OpenTelemetry and Semantic Conventions.  Today they interact with "process-container" for the data being generated.  Their tool understands what process-container is, and uses it for identity.
  - System Y defines a new way of modelling infrastructure.  They use containers and processes, but require a new type of identity that User X desires.
  - User X is forced to update their entity identity in Resource to be "Process-Container-System-Y".
     - They can no longer leverage OOTB entity detectionf or process or containers
     - System Y is forced to duplicate all behavior around process/container look up, even if it could be shared.
     - OpenTelemetry needs to be aware that Process-Container-System-Y and Process-Container are basically overlapping things and MAY reference the same item.
     
     
In other words, I think we've been having a thorough debate, and the consensus at the Entities WG was multiple entity associations. I think more design and outlining is needed to justify how a single entity paradigm would work in an open ecosystem, and how to address the challenges outliend.

Reply:
Wouldn't entities be used to encode different abstractions like service offerings or similar? Then we would immediately have orthogonal associated entities. If such abstractions can be subject to migrations, we would even have n:m relations between the physical representation modelled in resources and the entities encoding the various orthogonal views.

Reply:
Services are effectively logical groupings we enforce.  I believe, already, you can have different logical groupings and therefore different associated entities with a Resource.

If look at the actual attributes added by resource detection at the bottom, you'll see I think we already have N:M relations between physical<->logical representation.  Previously, the discussion was whether have a 1:1 Resource<->Entity relationship was ideal.  Given, at a minimum, we have 1:1 logical<->physical entity relationships, and we can have multiple logical models, so N:1, We're basically loosening this view.

There's a lot of rational and reasoning in the rest of the document on this.